Add `read_consistency` parameter to the APIs (#1371)

ffuugoo opened the pull request:
Add `read_consistency` parameter to the HTTP and gRPC APIs.

**TODO:**
- [x] add/update documentation

ffuugoo commented on src/actix/api/retrieve_api.rs:
```diff
@@ -52,7 +55,7 @@ pub async fn get_point(
     };
 
-    let response = do_get_point(toc.get_ref(), &collection_name, point_id).await;
+    let response = do_get_point(toc.get_ref(), &collection_name, point_id, todo!()).await;
```
Should we add an optional query parameter `read_consistency` to the `get_point` API?

generall replied:
yes, why not?

ffuugoo replied:
Done.

ffuugoo commented on lib/collection/src/recommendations.rs:
```diff
@@ -69,6 +70,7 @@ async fn retrieve_points(
                 ids,
                 with_payload: Some(WithPayloadInterface::Bool(false)),
                 with_vector: WithVector::Selector(vector_names),
+                read_consistency: ReadConsistency::Factor(1), // TODO: Which `ReadConsistency` should be used here!?
```
See the `TODO`.

Should we propagate the read consistency value (from `recommend_batch_by`) used by the respective recommend request in the batch here?

ffuugoo replied:
Decided to redesign requests/messages a bit. The question is resolved with the redesign.

ffuugoo commented on lib/collection/src/collection.rs:
```diff
@@ -902,6 +903,7 @@ impl Collection {
             ids: search_result.iter().map(|x| x.id).collect(),
             with_payload,
             with_vector,
+            read_consistency: ReadConsistency::Factor(1), // TODO: Which `ReadConsistency` should be used here!?
```
See the `TODO`.

Seems like we should simply propagate read consistency (from the `search_batch` method) used by respective search request in the batch, but I'd better clarify just in case.

ffuugoo replied:
Ditto.

ffuugoo reviewed (commented):
@generall I need a bit of help here.

ffuugoo commented:
OMG, the PR grew to 500 lines, just as `WriteOrdering` one did. 🙈

ffuugoo commented:
For some reason `ReadConsistency::Factor` deserialization from query parameter does not work in HTTP API. :/

E.g.:
- `curl 'http://127.0.0.1:6333/collections/test_collection/points/1?read_consistency=all'` works just fine
- but `curl 'http://127.0.0.1:6333/collections/test_collection/points/1?read_consistency=2'` returns `Query deserialize error: data did not match any variant of untagged enum ReadConsistency` error
- all while the test in `consistency_params.rs` pass just fine

I'll figure it out tomorrow.

Other than this issue, I only need to run some local tests in clustered mode, and it's done.

ack @generall

agourlay commented on lib/api/src/grpc/qdrant.rs:
```diff
@@ -2397,6 +2434,29 @@ impl WriteOrderingType {
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, ::prost::Enumeration)]
 #[repr(i32)]
+pub enum ReadConsistencyType {
+    /// Send request to all nodes and return points which present on all of them
```
```suggestion
    /// Send request to all nodes and return points which are present on all of them
```

same suggestion on all variants :)

agourlay commented on lib/collection/src/operations/conversions.rs:
```diff
@@ -585,6 +585,7 @@ impl<'a> From<CollectionSearchRequest<'a>> for api::grpc::qdrant::SearchPoints {
                 DEFAULT_VECTOR_NAME => None,
                 vector_name => Some(vector_name.to_string()),
             },
+            read_consistency: None,
```
why is there no param. no forward in this conversion?

ffuugoo replied:
It's used for `RemoteShard` only, and we don't propagate read consistency over internal API (or everything explode). There are safeguards for this in the internal API service definition also, but kinda makes sense to set `None` here as well.

ffuugoo commented:
@generall Done. Fixed the issue with deserialization, fixed the bug with payload comparison, rebased on latest `dev`. I've also run a #1408 test with read consistency enabled. I've had to add additional sleep after killing uploader processes (see my comment in the #1408), however, with the read consistency enabled the test pass 10 out of 10 times, while it fails a few times without the feature.

agourlay commented on lib/collection/src/operations/consistency_params.rs:
```diff
@@ -53,23 +32,150 @@ impl Default for ReadConsistency {
+        })?;
+
+        let consistency = match value {
+            read_consistency::Value::Factor(factor) => Self::Factor(factor.try_into().unwrap()),
```
is it possible to remove the `unwrap` with `try_into()?` here?

ffuugoo replied:
Sure. I'll fix it.

agourlay replied:
great, there is another one in the diff I believe :)

ffuugoo replied:
It's in the `From` implementation, so a bit more annoying to fix.

ffuugoo commented:
@generall A single character typo (that I've introduced myself, of course) caused The Bug. Should have found it sooner, as everything was hinting me in the right direction... but, well, most bugs seem _obvious_ in retrospect. :/

generall commented on lib/segment/src/types.rs:
```diff
@@ -523,6 +524,16 @@ impl From<Map<String, Value>> for Payload {
     }
 }
 
+pub fn serialize_payload<S>(point: &Option<Payload>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    point
+        .as_ref()
+        .filter(|payload| !payload.is_empty())
+        .serialize(serializer)
+}
```
Disagreed.
Payload should not be null if user uploaded `{}` as payload. 
Overall, I think, payload should only be `null` if it is explicitly disabled in request: `"with_payload": false`

ffuugoo replied:
Agree. I was trying to "cheat" this `{}` vs. `null` errors, but then realized it's a dumb idea. I'll fix this properly tomorrow.

However, now that you mentioned that user can upload `{}` explicitly... we do not distinguish between "no payload" and "empty payload" in the gRPC API (which means we do not distinguish them in the internal API as well), so how can we tell one from the other? 🤔

Seems like the best we can do is to ensure we return `null` if `with_payload: false`, and `{}` if `with_payload: true`.

ffuugoo replied:
It's impossible to add an `optional` attribute to a `map` field in ProtoBuf, so I had to change `Record` and `ScoredPoint` conversions from their corresponding proto-messages.

ffuugoo commented:
@generall Can you, please, take a final look at the PR?
